### ISSUE TITLE
[RAC] SelectValue render state

### ISSUE THREAD

**Issue by blue-infinex** [adobe/react-spectrum#7458]
### Provide a general summary of the issue here

The `ListBoxItem` render function is missing `isDisabled` when called inside `SelectValue`.

### 🤔 Expected Behavior?

The render fn of a `ListBoxItem` should receive the parent state when rendered within `SelectValue`.

### 😯 Current Behavior

The `ListBoxItem` render function is missing `isDisabled` when called inside `SelectValue`.

### 💁 Possible Solution

Pass the parental `isDisabled` state down [from `SelectValue`](https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/src/Select.tsx#L251):

```diff
 let state = useContext(SelectStateContext)
 ...
 rendered = fn({
   ...
-  isDisabled: false,
+  isDisabled: Boolean(state.isDisabled),
   selectionMode: 'single',
   selectionBehavior: 'toggle'
 })
```

### 🖥️ Steps to Reproduce

Pass a render fn to a `ListBoxItem` and try to access state inside `SelectValue`.

### Version

1.4.1

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Microsoft Edge

### What operating system are you using?

macOS 14.6.1

**Comment by snowystinger** [comment]
The `SelectValue` isn't really something that cares about disabled since it's non-interactive, instead, the `Button` that you should render the `SelectValue` inside of has the `isDisabled` state.
See the styling section here https://react-spectrum.adobe.com/react-aria/Select.html#disabled

Let us know if I've missed the intent of this issue here.

**Comment by blue-infinex** [comment]
Yeah, fair enough. Sorry, I should've included more information about the problem rather than a desired solution.

**Context**

I've wrapped the RAC `ListBoxItem` creating an opinionated `ListBoxItemPrimitive` component, which influences things like icons and images that may be composed within an item. The primitive component is then used by both our `Select` and `Combobox`.

**Usage**

When an item is disabled via `disabledKeys` everything behaves as expected since all state is provided as render props. However, if a value is defaulted or selected before the `Select` becomes disabled that `isDisabled` state isn't passed along from `SelectValue`.

---

Let me know if I'm "holding it wrong" or something. I've supplemented the expected value with internal field context, it just seemed odd that the list box item's children are used in both instances but the render props aren't compatible.

**Comment by snowystinger** [comment]
Maybe an example in Stackblitz or codesandbox would be helpful.

To clarify a little more, you're using composeRenderProps to allow the same `children` signature on the ListBoxItemPrimitive that we have on ListBoxItem?
If so, then because your wrapper is opinionated, maybe it's better to restrict the children? You could use our implementation for ReactSpectrum as a guide https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/s2/src/Picker.tsx#L418

Otherwise, I think it's probably just a case we didn't come across and so missed, but it'd be good to confirm a little more.
Thanks for understanding.

**Comment by blue-infinex** [comment]
> […]you're using composeRenderProps to allow the same children signature on the ListBoxItemPrimitive that we have on ListBoxItem? If so, then because your wrapper is opinionated, maybe it's better to restrict the children?

Similar to the `PickerItem`, I'm restricting `children` to `ReactNode` and wrapping with context providers. Reduced code:

```jsx
<ListBoxItem>
  {(state) => (
    <>
      <Provider
        values={[
          [ImageContext, {
            className: cn(commonAdornmentStyles, {
              // Not sure if we need field context here, I'd expect the disabled state to be provided by `SelectValue`.
              "opacity-50 saturate-0": state.isDisabled || fieldCtx.isDisabled
            }),
          }],
        ]}
      >
        {props.children}
      </Provider>

      <SelectionIndicator />
    </>
  )}
</ListBoxItem>
```

The `DefaultProvider` is an interesting approach. I'll spend some more time looking through that file, but it seems like this might be a pretty rare requirement.

**Comment by snowystinger** [comment]
I'm going to close this one for now as there are ways to accomplish this right now, so it's not a priority for us. I think we could consider a contribution, in which case I can reopen this. Thanks for the discussion.
